### PR TITLE
feat(ids-admin): Publish permission

### PR DESCRIPTION
**Pull request by snaerseljan**
# Publish permission

## What

This PR adds Publish permission feature to our permission/scope.
Now we can easily duplicate environment with one click just like in our client logic.
This is now a modal instead of a modal route. I have updated the PublishClient also to be just a modal instead of a modal route. The reason for the modal route change is that the publish route was giving us unexpected behaviour when navigating back and forward.
 
## Why

It will easy our life easier when creating new env in permission

## TODO

- [x] Fix tests

## Screenshots / Gifs

image

And if error occurs

image

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: apps/services/auth/admin-api/src/app/v2/scopes/test/me-scopes.spec.ts [hunk 1]
```diff
@@ -545,19 +561,23 @@ describe('MeScopesController', () => {
...
         expect(response.status).toEqual(testCase.expected.status)
-
         if (response.status === 200) {
           // Assert response
           expect(response.body).toStrictEqual({
             ...testCase.expected.body,
+            grantToAuthenticatedUser: testCase.expected.body
+              .grantToAuthenticatedUser
+              ? testCase.expected.body.grantToAuthenticatedUser
+              : true,
```
Comment (eirikurn) [inline apps/services/auth/admin-api/src/app/v2/scopes/test/me-scopes.spec.ts:571]: This always assigns true.

File: libs/api/domains/auth-admin/src/lib/scope/dto/publish-scope.input.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { Environment } from '@island.is/shared/types'
+import { Field, InputType } from '@nestjs/graphql'
+
+@InputType('AuthAdminPublishScopeInput')
+export class PublishScopeInput {
+  @Field(() => Environment, { nullable: false })
```
Comment (eirikurn) [inline libs/api/domains/auth-admin/src/lib/scope/dto/publish-scope.input.ts:6]:
```suggestion
  @Field(() => Environment)
```

Nullable false is the default. Same for below.

File: libs/api/domains/auth-admin/src/lib/scope/scope.service.ts [hunk 3]
```diff
@@ -84,6 +95,51 @@ export class ScopeService extends MultiEnvironmentService {
...
+    // Fetch the scope from source environment
+    const sourceInput = await this.adminApiByEnvironmentWithAuth(
+      input.sourceEnvironment,
+      user,
+    )
+      ?.meScopesControllerFindByTenantIdAndScopeName({
+        tenantId: input.tenantId,
+        scopeName: input.scopeName,
+      })
+      .catch((error) => this.handleError(error, input.sourceEnvironment))
```
Comment (eirikurn) [inline libs/api/domains/auth-admin/src/lib/scope/scope.service.ts:116]:
```suggestion
      ?.catch((error) => this.handleError(error, input.sourceEnvironment))
```
Will this throw if `adminApiByEnvironmentWithAuth` returns undefined?

File: libs/auth-api-lib/src/lib/resources/admin/admin-scope.service.ts [hunk 4]
```diff
@@ -113,10 +113,34 @@ export class AdminScopeService {
       throw new BadRequestException(`Scope name "${input.name}" already exists`)
     }
 
+    // We need to manually validate displayName and description, since we reuse the same dto for both create and publish
+    const translatedValuesErrorMsg =
+      'Scope displayName and description are required'
+
+    if (!input.displayName || !input.description) {
+      throw new BadRequestException(translatedValuesErrorMsg)
+    }
```
Comment (eirikurn) [inline libs/auth-api-lib/src/lib/resources/admin/admin-scope.service.ts:122]: Don't understand why this isn't validated like other fields?

File: libs/auth-api-lib/src/lib/resources/admin/admin-scope.service.ts [hunk 5]
```diff
@@ -113,10 +113,34 @@ export class AdminScopeService {
       throw new BadRequestException(`Scope name "${input.name}" already exists`)
     }
 
+    // We need to manually validate displayName and description, since we reuse the same dto for both create and publish
```
Comment (eirikurn) [inline libs/auth-api-lib/src/lib/resources/admin/admin-scope.service.ts:116]: This API doesn't have any concept of "publish". This PR is reusing the DTO for post (create) and patch (update).

File: apps/services/auth/admin-api/src/app/v2/scopes/test/me-scopes.spec.ts [hunk 6]
```diff
@@ -265,21 +260,42 @@ const createTestCases: Record<string, CreateTestCase> = {
       body: expectedCreateOutput,
     },
   },
+  'should publish scope and have access as current user': {
```
Comment (eirikurn) [inline apps/services/auth/admin-api/src/app/v2/scopes/test/me-scopes.spec.ts:263]: Maybe rephrase to explain expanded create/post functionality, since the rest API has no concept of publish.

File: libs/portals/admin/ids-admin/src/screens/Client/PublishClient/PublishClient.tsx [hunk 7]
```diff
@@ -1,130 +1,116 @@
...
+              targetEnvironment: data.targetEnvironment,
+              sourceEnvironment: data.sourceEnvironment,
+            },
+          },
+        })
+      } catch (e) {
+        setError(true)
+      }
+    },
+    [publishClient, tenantId],
```
Comment (eirikurn) [inline libs/portals/admin/ids-admin/src/screens/Client/PublishClient/PublishClient.tsx:64]: Check these lint warnings?

**Review by eirikurn — COMMENTED** [review]
Great job! 🎉Focused on reviewing the backend code and just skimmed the frontend. Happy to pair review specific frontend code if you want.

**Review by eirikurn — APPROVED** [review]
Approving since the comments were all nice to have.

File: libs/api/domains/auth-admin/src/lib/scope/scope.service.ts (hunk 3 above)
Comment (snaerseljan) [reply libs/api/domains/auth-admin/src/lib/scope/scope.service.ts:116]: I am pretty sure it does not, at least this does not throw

```
const p = new Promise((resolve) => resolve(undefined))
await p.catch((err) => {
    throw err
})
```

or am I missing or misunderstanding something

File: apps/services/auth/admin-api/src/app/v2/scopes/test/me-scopes.spec.ts (hunk 1 above)
Comment (snaerseljan) [reply apps/services/auth/admin-api/src/app/v2/scopes/test/me-scopes.spec.ts:571]: I will add isDefined check there

File: libs/auth-api-lib/src/lib/resources/admin/admin-scope.service.ts (hunk 4 above)
Comment (snaerseljan) [reply libs/auth-api-lib/src/lib/resources/admin/admin-scope.service.ts:122]: The fields are optional code wise because we extend the AdminPatchScopeDto, but required in create in the real world.

```
export class AdminCreateScopeDto extends AdminPatchScopeDto {
  @IsString()
  @IsNotEmpty()
  @ApiProperty({ example: '@island.is' })
  name!: string
}
```

File: libs/auth-api-lib/src/lib/resources/admin/admin-scope.service.ts (hunk 5 above)
Comment (snaerseljan) [reply libs/auth-api-lib/src/lib/resources/admin/admin-scope.service.ts]: Removed comment
